[Tensorpipe Agent] Timeouts for RPC requests

osalpekar opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #38589 Handling Active Call Count through Future Callback
* **#38448 [Tensorpipe Agent] Timeouts for RPC requests**

This PR implements timeout support for RPCs, and respects the new per-RPC timeout functionality.

A map containing RPC futures, keyed by an expiration time, is populated by the send function for each RPC.

A separate watchdog thread polls this map and sets all incomplete futures with errors.
Note: we cannot set errors to a future with the lock held (this will trigger callbacks immediately and, if one of the callback functions tries to acquire the lock that we held when setting the error, we will have a lock order cycle). Thus we add all incomplete futures to a list, and then iterate through the list outside the lock to set errors on those futures if necessary.

Fixes: https://github.com/pytorch/pytorch/issues/33980

Differential Revision: [D21468526](https://our.internmc.facebook.com/intern/diff/D21468526/)

osalpekar committed:
[Tensorpipe Agent] Timeouts for RPC requests

This PR implements timeout support for RPCs, and respects the new per-RPC timeout functionality.

A map containing RPC futures, keyed by an expiration time, is populated by the send function for each RPC.

A separate watchdog thread polls this map and sets all incomplete futures with errors.
Note: we cannot set errors to a future with the lock held (this will trigger callbacks immediately and, if one of the callback functions tries to acquire the lock that we held when setting the error, we have a lock order cycle). Thus we add all incomplete futures to a list, and then iterate through the list outside the lock to set errors on those futures if necessary.

Differential Revision: [D21468526](https://our.internmc.facebook.com/intern/diff/D21468526/)

[ghstack-poisoned]

lw commented on torch/csrc/distributed/rpc/tensorpipe_agent.h:
```diff
@@ -159,6 +162,31 @@ class TensorPipeAgent : public RpcAgent {
   uint64_t nextMessageID_{0};
 
+  // Map to store the expiration times for each message.
+  std::map<steady_clock_time_point, std::vector<std::shared_ptr<FutureMessage>>>
+      timeoutMap_;
```
Is there a specific reason to use a `std::map<TKey, std::vector<TValue>>` rather than a `std::multimap<TKey, TValue>`? I guess dealing with an entire vector at a time may be more efficient than dealing with multiple futures over multiple iterations, but we also don't really expect to have that many futures with _exactly_ the same timeout, down to the same millisecond. And dealing with one future at a time rather than a whole vector of them would simplify the above code further.

jjlilley commented on torch/csrc/distributed/rpc/tensorpipe_agent.cpp:
```diff
@@ -25,6 +25,9 @@ namespace distributed {
 
 constexpr long kToMilliseconds = 1000;
+// This large time duration is for the timeoutMapCV_. We cannot use
+// std::chrono::time_point::max() due to a known overflow-related bug.
```
Luca, I don't have the link, but we noticed in process_group changes that when std::chrono::time_point::max() got converted into the appropriate timespec struct for the underlying pthread_cond_timedwait() syscall, the timepoint value would overflow.

It could work if we used a smaller max value, or just use the non-timedwait version.

lw replied:
Indeed in the processgroup agent we do either `wait` or `wait_until` depending on whether the timeout is infinite:
https://github.com/pytorch/pytorch/blob/15da26f8aafe50fb34d343d872c7461546d7a317/torch/csrc/distributed/rpc/process_group_agent.cpp#L783-L788

Could we at least explain the issue better in the comment rather than expecting the user to know it already?

lw commented on torch/csrc/distributed/rpc/tensorpipe_agent.cpp:
```diff
@@ -402,6 +427,73 @@ std::shared_ptr<FutureMessage> TensorPipeAgent::send(
+
+    lock.unlock();
+
+    // Set an error on futures added to the errorFutures map. We do this
+    // outside the lock to prevent potential lock-order-inversions by callbacks
+    // triggered by the serError call.
+    for (const auto& future : errorFutures) {
+      std::string errorMsg = c10::str(
+          "RPC ran for more than set timeout and will now be marked with an error");
+      // Using setErrorIfNeeded so we don't set an error if the future was
+      // completed since adding to the errorFutures vector.
+      future->setErrorIfNeeded(errorMsg);
+    }
+    errorFutures.clear();
```
Another concern: when we time out a future we should decrease the count of active calls, however by using `setErrorIfNeeded` we don't know whether we actually did time it our or not.

I think a robust solution would be to add a `then` callback to the future as soon as the agent creates it (in the `send` method) and decrease the count in there. But I'm not sure of how PyTorch's futures work and whether this actually works as I imagine...

osalpekar replied:
Yes we definitely need to decrease the active callcount. For now, we can just check if the future has an error (using `future->hasError`)  after calling `setErrorIfNeeded`.

osalpekar committed:
Update on "[Tensorpipe Agent] Timeouts for RPC requests"


This PR implements timeout support for RPCs, and respects the new per-RPC timeout functionality.

A map containing RPC futures, keyed by an expiration time, is populated by the send function for each RPC.

A separate watchdog thread polls this map and sets all incomplete futures with errors.
Note: we cannot set errors to a future with the lock held (this will trigger callbacks immediately and, if one of the callback functions tries to acquire the lock that we held when setting the error, we will have a lock order cycle). Thus we add all incomplete futures to a list, and then iterate through the list outside the lock to set errors on those futures if necessary.

Fixes: https://github.com/pytorch/pytorch/issues/33980

Differential Revision: [D21468526](https://our.internmc.facebook.com/intern/diff/D21468526/)

[ghstack-poisoned]

lw commented on torch/csrc/distributed/rpc/tensorpipe_agent.cpp:
```diff
@@ -402,6 +429,58 @@ std::shared_ptr<FutureMessage> TensorPipeAgent::send(
+        --clientActiveCalls_;
+      }
+    }
+    timedOutFutures.clear();
```
Actually, no need to explicitly clear it: the vector will be automatically destroyed as soon as we leave the scope in the next line.

lw commented on torch/csrc/distributed/rpc/tensorpipe_agent.cpp:
```diff
@@ -402,6 +429,58 @@ std::shared_ptr<FutureMessage> TensorPipeAgent::send(
+          "RPC ran for more than set timeout and will now be marked with an error");
+      // Using setErrorIfNeeded so completed futures are ignored.
+      future->setErrorIfNeeded(errorMsg);
+      if (future->hasError()) {
+        // We only want to decrease the active call count if the future has an
+        // error. If it was successful, the counter was already updated in the
+        // send() function above.
+        --clientActiveCalls_;
+      }
```
I still think this could lead to wrong counts: if the future had already been marked with an error (say, because of an I/O problem), then the count had already been decreased, and yet here we would decrease it a second time.

I see several ways to fix this:
- change the `setErrorIfNeeded` method of the torch::utils::Future to return a bool indicating whether the error was actually set (or whether the future was already complete)
- As soon as we create the Future in the send method, we call `addCallback` on it with a lambda that decreases `clientActiveCalls_`. The future ensures that such callbacks are only called once, the first time the future is marked complete (whether successfully or with an error), and therefore guarantees no double-decrease
- We could change the other code paths where we complete a future (also both for success and error outcomes) so that they remove the future from the `timeoutMap_`

I think the second solution is by far the simpler and more robust one.

osalpekar replied:
Implemented option 2 in the next PR in this stack (to modularize the changes a bit).

lw replied:
Thanks, looks good! And, as always, it's good to split up changes as much as it's reasonable. One tip though: when exposing a bug while writing a diff, I'd tend to fix it by _prepending_ a diff to the stack, rather than appending it. This way each commit is bug-free, and this is a good invariant to preserve, as it allows for example a bisect to run smoothly. In this case the bug introduced by this diff is minor (just a count that may sometimes be off), so it's fine.

lw commented on torch/csrc/distributed/rpc/tensorpipe_agent.cpp:
```diff
@@ -402,6 +429,58 @@ std::shared_ptr<FutureMessage> TensorPipeAgent::send(
+      if (std::chrono::steady_clock::now() >= earliestTimeout) {
+        break;
+      }
+      if (!timeoutMap_.empty()) {
```
```suggestion
      steady_clock_time_point earliestTimeout =
          std::chrono::steady_clock::now() + kLargeTimeDuration;
      if (!timeoutMap_.empty()) {
```
You could consider moving the definition of `earliestTimeout` inside the loop, as it's not used outside of it.

osalpekar committed:
Update on "[Tensorpipe Agent] Timeouts for RPC requests"


This PR implements timeout support for RPCs, and respects the new per-RPC timeout functionality.

A map containing RPC futures, keyed by an expiration time, is populated by the send function for each RPC.

A separate watchdog thread polls this map and sets all incomplete futures with errors.
Note: we cannot set errors to a future with the lock held (this will trigger callbacks immediately and, if one of the callback functions tries to acquire the lock that we held when setting the error, we will have a lock order cycle). Thus we add all incomplete futures to a list, and then iterate through the list outside the lock to set errors on those futures if necessary.

Fixes: https://github.com/pytorch/pytorch/issues/33980

Differential Revision: [D21468526](https://our.internmc.facebook.com/intern/diff/D21468526/)

[ghstack-poisoned]

osalpekar committed:
Update on "[Tensorpipe Agent] Timeouts for RPC requests"


This PR implements timeout support for RPCs, and respects the new per-RPC timeout functionality.

A map containing RPC futures, keyed by an expiration time, is populated by the send function for each RPC.

A separate watchdog thread polls this map and sets all incomplete futures with errors.
Note: we cannot set errors to a future with the lock held (this will trigger callbacks immediately and, if one of the callback functions tries to acquire the lock that we held when setting the error, we will have a lock order cycle). Thus we add all incomplete futures to a list, and then iterate through the list outside the lock to set errors on those futures if necessary.

Fixes: https://github.com/pytorch/pytorch/issues/33980

Differential Revision: [D21468526](https://our.internmc.facebook.com/intern/diff/D21468526/)

[ghstack-poisoned]
